capture/src/billing_limits.rs

```diff
@@ -0,0 +1,181 @@
+
+            // The goal is for the _entire_ iteration of this loop to take UPDATE_INTERVAL_SECS
+            // If our execution takes less time than that, the interval will pad it out.
+            loop {
+                let _span = tracing::debug_span!("billing limit update tick");
+                interval.tick().await;
+
+                let tokens = Self::fetch_limited(&client).await;
+
+                match tokens {
+                    Ok(tokens) => {
+                        tracing::debug!("{} limited tokens fetched from redis", tokens.len());
+
+                        // RAII guard on the write lock
+                        let mut l = limited.write().await;
+                        
+                        for i in tokens {
+                            l.insert(i);
+                        }
+                    }
+                    Err(e) => {
+                        // Oh no! Clear all limits. Something has gone wrong so let's play it safe.
+                        tracing::error!("error fetching limited set from redis: {}", e);
+
+                        let mut l = limited.write().await;
+                        l.clear();
+                    }
+                }
+            }
+        });
+
+        Ok(())
+    }
+
+    pub async fn is_limited(&self, key: &str) -> bool {
+        // RAII guard on read
+        // The RwLock allows many readers, but just one writer.
+        // This will yield the task if the write lock is currently held.
+        let l = self.limited.read().await;
+
+        l.contains(key)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::billing_limits::BillingLimiter;
+
+    // Test that a token _not_ limited has no restriction applied
+    // Avoid messing up and accidentally limiting everyone
+    #[tokio::test]
+    async fn test_not_limited() {
+        let limiter = BillingLimiter::new(vec![], None).expect("Failed to create billing limiter");
+        limiter
+            .start_updater()
+            .await
+            .expect("Failed to start updater");
+
+        assert_eq!(limiter.is_limited("idk it doesn't matter").await, false);
+    }
+
+    // Test that a token _not_ limited has no restriction applied
+    // Avoid messing up and accidentally limiting everyone
+    #[tokio::test]
+    async fn test_fixed_limited() {
+        let limiter = BillingLimiter::new(
+            vec![],
+            Some(
+                vec![String::from("some_org_hit_limits")]
+                    .into_iter()
+                    .collect(),
+            ),
+        )
+        .expect("Failed to create billing limiter");
+
+        limiter
+            .start_updater()
+            .await
+            .expect("Failed to start updater");
+
+        assert_eq!(limiter.is_limited("idk it doesn't matter").await, false);
+        assert!(limiter.is_limited("some_org_hit_limits").await);
+    }
+
+    // TODO: test the redis stuff somehow
```

Comment:
honestly I'm probably just going to wrap the client in my own trait, so I can inject failures and make sure we handle them right

Reply:
I really wanted to use `ConnectionLike` but it got kinda weird